### PR TITLE
[DI] Start `Catalog` frontend packages migration

### PR DESCRIPTION
**Pull request by camilaibs**
## Hey, I just made a Pull Request!

Start the `Catalog` plugin migration to the declarative integration format.
> [!Important]
> Skipping entity pages in the initial attempt. Mostly sidebar items, index page, and filter extensions.

### Known limitations

- The `Catalog` table will not be configurable initially, but we are open to suggestions about what could be configurable;
- The `CatalogEntityPage` will be migrated in a follow-up pull request.

### Usage example

> [!Note]
> This is experimental, **documentation will be updated in a follow-up pull request**.

Only one step is required to start using the `Catalog` plugin in Backstage, which is to install the `@backstage/plugin-catalog` package in a Backstage app that supports declarative integration (e.g.,  [`app-next`](https://github.com/backstage/backstage/tree/master/packages/app-next )).

```sh
yarn add @backstage/plugin-catalog
```

The experimental `Catalog` plugin comes with a default implementation for the `CatalogApi` and `StarredEntitiesApi`; also includes a default `CatalogIndexPage` as well a few `Catalog*Filter` extensions. All of this is configurable via `app-config.yaml` file, for example:

```yaml
# app-config.yaml
app:
  experimental:
    packages: 'all' # ✨

  extensions:
   # customize catalog index page path
    - catalog:
         config:
           path: '/entities'
   # customize catalog sidebar item label
    - catalog.nav.index:
        config:
          label: 'Entities'
    # disable default catalog entity tag filter
    - catalog.filter.entity.tag: false
    # define initial catalog entity kind filter
    - catalog.filter.entity.kind:
        config:
          initialFilter: 'api'
```

### Local testing:

First, install the branch dependencies and start both frontend and backend declarative applications:

```sh
# Installing branch dependencies
yarn install

# Starting the backend application
cd packages/backend-next
yarn start

# Starting the frontend application
cd packages/app-next
yarn start
```

Finally, navigate to http://locahost:3000 and click on the `Catalog` sidebar item:

https://github.com/backstage/backstage/assets/6290749/809baa65-b3e5-40bd-9c02-701098fb66c9

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [inline plugins/catalog/src/alpha.tsx:244]: > [!Note]
> This is required to avoid the following error: "No path for routeRef{type=absolute,id=catalog:entity}"

File: plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx
Comment (camilaibs) [inline plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx:106]: > [!Note]
> Open for ideas on catalog table setup, not in this pull request scope but could definitely be a follow up.

File: packages/frontend-app-api/src/wiring/createApp.tsx
Comment (camilaibs) [inline packages/frontend-app-api/src/wiring/createApp.tsx:481]: > [!Note]
> The `CatalogIndexPage` depends on this API reference.

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [inline plugins/catalog/src/alpha.tsx:235]: Would that be interesting to be able to change the filter order via configuration?

File: plugins/catalog/alpha-api-report.md
Comment (Rugvip) [inline plugins/catalog/alpha-api-report.md:57]: General comment not directly related to this PR: I think this export pattern would be very confusing in the long run, plenty of overlap with React components. I think the only way around it long term is if we started to move all component exports to something like `/components`.

What I'm leaning towards for now is to not export any extensions at all, it's not like you can do much with them at this point anyway 😅. Main reason for now would probably be testing, but I would find it a bit unfortunate if packages where cherry-picking extensions from a different plugin, and besides, we could just disable extensions by default from a plugins and you can pick the ones you want that way. By removing the exports we can start seeing if there is actually a need to reference extensions directly at all.

File: plugins/catalog/api-report.md
Comment (Rugvip) [inline plugins/catalog/api-report.md:218]: Hmm, thoughts on making this an internal API for now? We've been resisting similar changes to this component because we want to be able to change the default catalog page quite freely. Then again this is at a pretty good level of abstraction though, thoughts @backstage/catalog-maintainers, @timbonicus?

File: plugins/catalog/src/alpha.tsx
Comment (Rugvip) [inline plugins/catalog/src/alpha.tsx:100]: I feel we need a different type here, since you can't really install any plain react element here and expect it to work. Maybe a just a plain table filter type that also forwards a react element, although a more elaborate refactor could be to investigate if it's possible to break the contract apart a bit more similar to scaffolder field extensions, but I am worried that'd get very messy very quickly.

File: plugins/catalog/src/alpha.tsx
Comment (Rugvip) [inline plugins/catalog/src/alpha.tsx:203]: Feeling that async loading of these is the way to go, but also gonna be a bit interesting to keep an eye on app perf as we start lazy loading more aggressively

File: plugins/catalog/src/alpha.tsx
Comment (Rugvip) [reply plugins/catalog/src/alpha.tsx:235]: Certainly a tricky one, it's possible right now, but you have to disable and then enable all filters again. Maybe we could try out a design where the catalog plugin can be configured to disable all filter extensions by default? That'd be the most convenient solution for changing the order I think

File: plugins/catalog/src/alpha.tsx
Comment (Rugvip) [inline plugins/catalog/src/alpha.tsx:249]: 👌

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice! 😁 

Finding this very useful for challenging the design and solving real world problems

**Comment by Rugvip** [comment]
For @backstage/catalog-maintainers, please excuse us running all over your lawn 😅. We're using a couple of different plugins to aid us in the design of the new declarative integration system. We'll most likely take a stab at entity pages in not too long too, let us know if you want to be involved in that early or would rather wait for the DI system to be more finished, with docs etc.

File: plugins/catalog/src/alpha.tsx
Comment (freben) [reply plugins/catalog/src/alpha.tsx:235]: Didn't we reorder extensions in such a way that explicitly mentioned ones come before those that weren't?

Comment (freben) [reply plugins/catalog/src/alpha.tsx:235]: Oh I see - it applies to previously disabled things only.

https://github.com/backstage/backstage/blob/master/packages/frontend-app-api/src/wiring/parameters.test.ts#L110

In theory this could be expanded to work for enabled things too right?

File: plugins/catalog/api-report.md
Comment (camilaibs) [reply plugins/catalog/api-report.md]: I extracted a base catalog page layout component to avoid changing the default catalog page props, does it look good to you 🙂 ? Here is the commit: https://github.com/backstage/backstage/pull/20355/commits/dfd0b8eec5d6ad190efdc9d2934a189f58f0d251.

File: plugins/catalog/alpha-api-report.md
Comment (camilaibs) [reply plugins/catalog/alpha-api-report.md:57]: That's a good call out! At first, I thought that having access to the `at` property would help in overriding the extensions to avoid using magic strings for the attachment point, but after your comment, I realized that we already have access to the extensions through the plugin instance, like:
```tsx
import catalogPlugin from '@backstage/plugin-catalog';
const fistExtension = catalogPlugin.extensions[0].id;
```
I'll remove the exports for now 🙌🏻

Comment (camilaibs) [reply plugins/catalog/alpha-api-report.md]: Here os the commit 🙂 : https://github.com/backstage/backstage/pull/20355/commits/579279f1746b12b6e0bd34db49b3cc46cd15d1e0

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx]: Thinking now, that maybe just the page should be async because the user might not access it on the first application load, but when the user accesses the page, the filters should be immediately there... I'll convert the filters to be synchronously loaded 🙂

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx:224]: I wonder if that's the case the filter layout component should support users reordering filters through the UI instead and save it in the config or user settings... Thoughts on this other possibility 🙂 ?

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx:100]: Could you please point me to the ScaffolderFields implementation file?

File: plugins/catalog/src/alpha.tsx
Comment (Rugvip) [reply plugins/catalog/src/alpha.tsx:224]: @freben not really, because then things would start getting reordered when all you want to do is configured them. Would likely need some extra flag like:

```yaml
app:
  extensions:
  - my.yaml.extension:
      reorder: true
```

At a glance I kinda hate it :p

Yep I do think that the answer in some cases might be to make it user configurable, not something we can rely on though imo since it's quite a bit of work

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx]: Done 🙌🏻 : https://github.com/backstage/backstage/pull/20355/commits/ecaf52aea5076fb5455e2d66c0070910b85867fc

Comment (Rugvip) [reply plugins/catalog/src/alpha.tsx]: Oh no please revert that 😁, this will cause the filters to be eagerly loaded. While the catalog page it lazy these extensions end up having an eager reference to the entire filter component

Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx]: I dropped the commit 😅

File: plugins/catalog/src/alpha.tsx
Comment (Rugvip) [reply plugins/catalog/src/alpha.tsx:100]: This kind of things: https://github.com/backstage/backstage/blob/cbed529916fa87ac0f1145d5f9a1340170dfb838/plugins/scaffolder-react/src/extensions/index.tsx#L39. The key there being that it's an element as well as additional configuration surrounding it. Tbh it's what we're already doing a lot in DI.

Thing is that we currently have a contract for how the field interacts with the outside through the react context that the parent page provides. This is a quite hidden interface and completely opaque to the core framework. If we instead lifted out the context and refactored it into output that the filter extensions provide then we'd just have a single place where the communication happens and a more explicit connection between the parent page and the filter extensions.

I don't think any of that is worth doing at this point thought and it may not be the best design either. Happy to leave this as a known thing and see where else we encounter communication through react context and at a broader scale decide what we want to do with that.

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [inline plugins/catalog/src/alpha.tsx:85]: @Rugvip , @freben should I extract this to the `@backstage/plugin-catalog-react`? So other plugins, like `TechDocs could use this function for exporting their specific filters.

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx:100]: Thanks for sharing, I agree with you, let's leave it as is and refactor later when we have more use cases 🙂

File: plugins/catalog/src/alpha.tsx
Comment (Rugvip) [reply plugins/catalog/src/alpha.tsx:85]: So far I think it's only been for creating custom filters in apps right? Let's not make the change unless it's already present in the existing system imo, but we could consider this in the future.

**Review by Rugvip — APPROVED** [review]
I think it's good to go 😁 🎉

**Comment by freben** [comment]
oof a conflict got in!

File: plugins/catalog/src/alpha.tsx
Comment (freben) [inline plugins/catalog/src/alpha.tsx:219]: This one is interesting. This lets people add any "component" (actually element) of any type here, even non-filters. We mentioned this in some meeting right? Not sure if we should worry about it or not.

This goes two ways. Even if we add some second discriminator data field, we still have the opposite problem that it gets possible to mount entity filters on generic reactElement inputs - where they will inevitably fail to render since there's no entity filter context around them.

Hm.

Should we add a

```ts
export const catalogExtensionData = {
  catalogFilter: createExtensionDataRef<{ element: JSX:Element }>('catalog.filter'),
};
```

in `@backstage/plugin-catalog-react`? Or is this just wasteful? @Rugvip

File: plugins/catalog/src/alpha.tsx
Comment (freben) [inline plugins/catalog/src/alpha.tsx:230]: see above

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx:214]: For this catalog example, I had to use the same id as the routeRef id...

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx:230]: For this other catalog page same situation, I had to use the same id as the routeRef id...

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx:219]: @Rugvip mentioned possibly using a type similar to scaffolder fields, interested to hear your thoughts as well @freben:
https://github.com/backstage/backstage/pull/20355#discussion_r1348830910,

File: plugins/catalog/src/alpha.tsx
Comment (freben) [reply plugins/catalog/src/alpha.tsx:214]: could we rename the ref?

File: plugins/catalog/src/alpha.tsx
Comment (freben) [reply plugins/catalog/src/alpha.tsx:219]: alright, so we leave it as reactElement for now and possibly revisit later, if i'm reading that right?

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx:214]: If I'm not mistaken, @Rugvip mentioned avoiding renaming the reference ID for now. Let me double chek. Are there any concerns with renaming the ref id @Rugvip ?

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx:219]: This is my understanding too, but let's double check wit @Rugvip 🙂

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx:235]: I'll update this once this pull request gets merged 🙌🏻 : https://github.com/backstage/backstage/pull/20574

File: plugins/catalog/src/alpha.tsx
Comment (camilaibs) [reply plugins/catalog/src/alpha.tsx:219]: I'll update this once this pull request gets merger 🙌🏻 : https://github.com/backstage/backstage/pull/20574

File: packages/app-next/src/App.tsx
Comment (camilaibs) [inline packages/app-next/src/App.tsx:81]: Created a separate pull request for exporting the TechDocs routes: https://github.com/backstage/backstage/pull/20605/files#diff-c34ee22bb6f58cec5c50bd549823f9e998e98aab7e3615abda60f22c3971e0c6R164

File: plugins/catalog/src/alpha.tsx
Comment (Rugvip) [reply plugins/catalog/src/alpha.tsx:224]: Yep I'm honestly even a bit torn on migrating this at this point. I feel the system has come far enough that we'd be able to build a proper abstraction around catalog filters instead. Migrating to an API that mimics the existing react-based APIs might be counter productive in that we're fairly certain we'll want to remove it in the future and we don't really need it right now.

**Review by Rugvip — APPROVED** [review]
Nice! Let's :shipit: so we can build on top of it to figure out the entity pages.

As mentioned this might not be what we want to do for the catalog page filters, but let's revisit in the future

**Pull request merged**
